### ISSUE TITLE
Assertion in `rbuffer_consumed_compact` tripped

### ISSUE THREAD

**Issue by RedBeard0531** [neovim/neovim#21299]
### Neovim version (nvim -v)

v0.9.0-dev-447+g6c285a6e49 (Also experienced crashes with v0.8, but may be different cause)

### Vim (not Nvim) behaves the same?

no, this function is nvim-only I think

### Operating system/version

ArchLinux

### Terminal name/version

konsole

### $TERM environment variable

xterm-256color

### Installation

Arch neovim-git (also experience crashes with official neovim v0.8 package)

### How to reproduce the issue

Not exactly sure how to repro. I've been experiencing neovim crashes lately. I don't know exactly what triggers it, but some file types (typescript and markdown) seem to trigger it more often than others. Sometimes it crashes soon after opening many times in a row, sometimes I can go many days between crashes. It seems to happen most often when saving, but sometimes it happens just when moving the cursor around.

I typically use neovim-gtk, but recently tried running plain neovim in a terminal to see if that helped. It still crashes, but now I get a helpful message out (although it leaves my terminal in a trashed state). 

```
nvim: /home/mstearn/.cache/yay/neovim-git/src/neovim-git/src/nvim/rbuffer.c:168: rbuffer_consumed_compact: Assertion `(size_t)(buf->read_ptr - buf->write_ptr) == buf->size' failed.
zsh: IOT instruction  nvim
```

The failing assert is on [this line](https://github.com/neovim/neovim/blob/4b02291f81b76eda029d4eae1c1d59c1a946a032/src/nvim/rbuffer.c#L168). That function is only called from [one place](https://github.com/neovim/neovim/blob/4b02291f81b76eda029d4eae1c1d59c1a946a032/src/nvim/msgpack_rpc/channel.c#L238), and both were introduced in [this commit](https://github.com/neovim/neovim/commit/d5f047bee04a42f40425c34061c84b39af846e1f). While I don't have a repro, I am hoping that something near one of those lines will jump out at someone who knows this code (possibly @bfredl who committed it) as obviously incorrect.

### Expected behavior

I'd really like my editor not to crash :)

### Actual behavior

neovim crashes

**Comment by alEXE-tech** [comment]
Also seem to be having this issue. Arch linux user as well downloading neovim nightly. My build script is as follows:
```
#!/bin/bash
cd ~
sudo rm -r neovim
git clone https://github.com/neovim/neovim
cd neovim
sudo make CMAKE_BUILD_TYPE=RelWithDebInfo
sudo make install
cd ../
sudo rm -r neovim
```

**Comment by emilienlemaire** [comment]
I have the same issue on macOs, after building it myself and with downloading the tarball.
The error is:
`Assertion failed: ((size_t)(buf->read_ptr - buf->write_ptr) == buf->size), function rbuffer_consumed_compact, file rbuffer.c, line 168.`

**Comment by emilienlemaire** [comment]
It seems to be triggered by a changed made between 4703e56 and c590641 but I cannot say more, I might try later to find the specific commit where this begins

**Comment by zeertzjq** [comment]
There are two different assertions mentioned here.
- `rbuffer.c:144: rbuffer_consumed: Assertion count && count <= buf->size' failed` can be simply triggered by `echo | nvim --clean --embed`, but `rbuffer_consumed` is used in many places.
- `rbuffer.c:168: rbuffer_consumed_compact: Assertion `(size_t)(buf->read_ptr - buf->write_ptr) == buf->size' failed` is the one first mentioned in the issue, and `rbuffer_consumed_compact` is only used in one place.

**Comment by emilienlemaire** [comment]
I've had both but I do not know when what triggers one or the other

**Comment by grimm26** [comment]
I get the ` rbuffer.c:144: rbuffer_consumed: Assertion count && count <= buf->size' failed..` one when deleting multiple lines at a time with something like `4dd` but not when I do one line at a time. Log seems to point to treesitter:
```
RR 2022-11-18T15:59:31.211 nvim.521393.0 decor_provider_invoke:47: error in provider treesitter/highlighter:line: Error executing lua: ...hedn/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:499: bad argument #2 to 'handler' (string expected, got nil)
stack traceback:
        [C]: in function 'handler'
        ...hedn/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:499: in function 'match_preds'
        ...hedn/usr/share/nvim/runtime/lua/vim/treesitter/query.lua:584: in function 'iter'
        ...sr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:195: in function 'fn'
        ...r/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:210: in function 'for_each_tree'
        ...sr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:168: in function 'on_line_impl'
        ...sr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:240: in function <...sr/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:234>
        [C]: in function 'redrawstatus'
        ...site/pack/packer/opt/noice.nvim/lua/noice/util/hacks.lua:267: in function <...site/pack/packer/opt/noice.nvim/lua/noice/util/hacks.lua:263>
ERR 2023-01-03T14:56:13.886 nvim.91933.0 os_set_cloexec:498: Failed to get flags on descriptor 3: Bad file descriptor
ERR 2023-01-03T15:09:26.994 nvim.94266.0 os_set_cloexec:498: Failed to get flags on descriptor 3: Bad file descriptor
ERR 2023-01-03T15:11:59.662 nvim.95548.0 os_set_cloexec:498: Failed to get flags on descriptor 3: Bad file descriptor
ERR 2023-01-03T15:25:33.307 nvim.100843.0 os_set_cloexec:498: Failed to get flags on descriptor 3: Bad file descriptor
ERR 2023-01-03T17:02:34.265 ?.126901   chan_close_with_error:573: RPC: failed to decode msgpack
ERR 2023-01-03T17:07:14.927 ?.132350   chan_close_with_error:573: RPC: failed to decode msgpack
ERR 2023-01-03T17:18:45.445 ?.139615   chan_close_with_error:573: RPC: failed to decode msgpack
```

**Comment by lewis6991** [comment]
Unless anyone can provide more useful information, can anyone else that experiences this just provide a 👍 reaction to the OP so we can gauge who this affects.

**Comment by lewis6991** [comment]
I think the likely culprits of this could be 6ba34e21fee2a81677e8261dfeaf24c8cd320500 (since it's LSP related which I think is the only user of `rbuffer_consumed_compact`) or cce736218f465511194465410e1ba23b5530e46f (since this was quite a big change).

**Comment by lewis6991** [comment]
I have a backtrace. No idea if it's useful or not:

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x19ac4e1b0 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x19ac84cec pthread_kill + 288
2   libsystem_c.dylib             	       0x19abbe2c8 abort + 180
3   libsystem_c.dylib             	       0x19abbd620 __assert_rtn + 272
4   nvim                          	       0x100448790 rbuffer_consumed_compact + 284
5   nvim                          	       0x1003d46a8 receive_msgpack + 344
6   nvim                          	       0x1002ddd68 read_cb + 340
7   nvim                          	       0x10011e470 uv__read + 328 (stream.c:1158) [inlined]
8   nvim                          	       0x10011e470 uv__stream_io + 768 (stream.c:1226)
9   nvim                          	       0x100123bd8 uv__io_poll + 1000 (kqueue.c:378)
10  nvim                          	       0x10011538c uv_run + 384 (core.c:406)
11  nvim                          	       0x1002db688 loop_poll_events + 284
12  nvim                          	       0x10051db64 ui_client_run + 1052
13  nvim                          	       0x1003865c8 main + 16272
14  dyld                          	       0x19a95be50 start + 2544
```

**Comment by zeertzjq** [comment]
After #18375 simply this can cause this assert failure:
```vim
:lua io.stdout:write('a')
```

This issue was created before #18375 though.

**Comment by bfredl** [comment]
yes, stdin/stdout fd:s will need to be renamed in the server. I'll look into a fix.

**Comment by mosheavni** [comment]
> After #18375 simply this can cause this assert failure:
> 
> ```viml
> :lua io.stdout:write('a')
> ```
> 
> This issue was created before #18375 was merged though.

I can confirm that this reproduces the error, after typing this command on a buffer, NeoVim crashes.

**Comment by zeertzjq** [comment]
This `assert((size_t)(buf->read_ptr - buf->write_ptr) == buf->size);` doesn't seem to make any sense. It can fail whenever `count != buf->size`.

**Comment by zeertzjq** [comment]
This assert failure can happen with the following sequence:
1. `rbuffer_produced()` is called with a `count` of 65536, then the buffer is full (`buf->size == 65536`) and `buf->read_ptr` and `buf->write_ptr` are both set to `buf->start_ptr`.
2. 65533 bytes are parsed as msgpack, so `rbuffer_consumed_compact()` is call with a `count` of 65533.
3. After consuming 65533 bytes, `buf->read_ptr == buf->start_ptr + 65533` and `buf->size == 3`, while `buf->write_ptr == 0`
4. Assert failure

**Comment by zeertzjq** [comment]
However, when rbuffer isn't full, `buf->write_ptr == buf->write_ptr + buf->size`, so this assertion failure whenever `buf->size != 0`.

**Comment by lewis6991** [comment]
With #21675 I'm now getting a different crash but In the same area of the code:

```
nvim(25902,0x1f5feba80) malloc: *** error for object 0x117814440: pointer being freed was not allocated
```

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x19ac4e1b0 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x19ac84cec pthread_kill + 288
2   libsystem_c.dylib             	       0x19abbe2c8 abort + 180
3   libsystem_malloc.dylib        	       0x19aadcb1c malloc_vreport + 908
4   libsystem_malloc.dylib        	       0x19aae0330 malloc_report + 64
5   libsystem_malloc.dylib        	       0x19aaece60 find_zone_and_free + 308
6   nvim                          	       0x1046864c0 rpc_send_event + 660
7   nvim                          	       0x1047bc2c0 tinput_flush + 480
8   nvim                          	       0x1047bb3c8 tinput_read_cb + 156
9   nvim                          	       0x104599bd0 read_cb + 340
10  nvim                          	       0x1043da4a0 uv__read + 328 (stream.c:1158) [inlined]
11  nvim                          	       0x1043da4a0 uv__stream_io + 768 (stream.c:1226)
12  nvim                          	       0x1043dfc08 uv__io_poll + 1000 (kqueue.c:378)
13  nvim                          	       0x1043d13bc uv_run + 384 (core.c:406)
14  nvim                          	       0x1045974c4 loop_poll_events + 284
15  nvim                          	       0x1047cf2c0 ui_client_run + 760
16  nvim                          	       0x104638708 main + 16496
17  dyld                          	       0x19a95be50 start + 2544
```

**Comment by matiasvc** [comment]
I have been getting this a lot lately. Any debugging steps I can do to help resolve this?

My neovim version:
```
❯ nvim --version
NVIM v0.9.0-dev
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-Se7AhW/neovim-0.9.0~ubuntu1+git202301110019-15ee93c0b-333b5866f=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g  -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fno-common -fdiagnostics-color=auto -fstack-protector-strong -DNVIM_UNIBI_HAS_VAR_FROM -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DMIN_LOG_LEVEL=3 -I/build/neovim-Se7AhW/neovim-0.9.0~ubuntu1+git202301110019-15ee93c0b-333b5866f/.deps/usr/include -I/usr/include -I/usr/include/luajit-2.1 -I/build/neovim-Se7AhW/neovim-0.9.0~ubuntu1+git202301110019-15ee93c0b-333b5866f/build/src/nvim/auto -I/build/neovim-Se7AhW/neovim-0.9.0~ubuntu1+git202301110019-15ee93c0b-333b5866f/build/include -I/build/neovim-Se7AhW/neovim-0.9.0~ubuntu1+git202301110019-15ee93c0b-333b5866f/build/cmake.config -I/build/neovim-Se7AhW/neovim-0.9.0~ubuntu1+git202301110019-15ee93c0b-333b5866f/src
Compiled by buildd@lcy02-amd64-001

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Comment by zeertzjq** [comment]
> This assert failure can happen with the following sequence:
> 
>     1. `rbuffer_produced()` is called with a `count` of 65536, then the buffer is full (`buf->size == 65536`) and `buf->read_ptr` and `buf->write_ptr` are both set to `buf->start_ptr`.
> 
>     2. 65533 bytes are parsed as msgpack, so `rbuffer_consumed_compact()` is called with a `count` of 65533.
> 
>     3. After consuming 65533 bytes, `buf->read_ptr == buf->start_ptr + 65533` and `buf->size == 3`, while `buf->write_ptr == 0`
> 
>     4. Assert failure

This was the case I found out using `rr` when I last got a crash with steps similar to the one mentioned in #21785.

**Comment by bfredl** [comment]
Ok, rbuffer needs to be refactored to actually behave like a ring buffer. As a quickfix, the only exception that could happen in msgpack code is when buffer is exactly full, which could be added to the assert.

**Comment by zeertzjq** [comment]
#21899 may fix some `rbuffer.c:144: rbuffer_consumed: Assertion count && count <= buf->size' failed` crashes.

**Comment by ianhomer** [comment]
Thanks for the fix @zeertzjq - I've updated to nightly and searching with telescope doesn't crash for me now. Previous nightly it would crash quite quickly.

**Comment by folke** [comment]
Not sure if this helps, but based on a number of user reports and by my own testing, it seems this issue only occurs on Macos. I can not reproduce this on Windows. The issue seems to be triggered by scrolling quickly in a buffer.

**Comment by FredeEB** [comment]
@folke I saw this on Ubuntu and Arch Linux as well

**Comment by folke** [comment]
@FredeEB with Noice? For Noice, this has been fixed in Noice and will be fixed upstream. But that was related to `vim.ui_attach`. There seems to be another issue that's not related to `vim.ui_attach` as well. And for me at least that issue only triggers on Macos

**Comment by lewis6991** [comment]
I think there were a few vectors that could trip this and we've fixed quite a few of them. So when reporting on this, make sure it is for the latest commit on master.

I haven't had this trip on me for a while now on MacOS, and I was getting it a lot before.

**Comment by lewis6991** [comment]
I haven't seen this for a full 2 days now so going to add the "needs response" label.

**Comment by folke** [comment]
ok, cool. Will test myself tomorrow on macos

**Comment by zeertzjq** [comment]
Closing as this is likely fixed.

**Comment by folke** [comment]
Yep, the noice issue is also resolved. Thank you all!

**Comment by jorgen99** [comment]
I just left a comment in #22199.

TL;DR 
Did a bisect from 0.7.2 and our crashes starts with commit d5f047bee04a42f40425c34061c84b39af846e1f

**Comment by agluck91** [comment]
I am on MacOs with latest MacOs, latest Alcritty terminal, latest NeoVim. The behavior that OP reported is happening for me now. It happens when either scrolling with mouse or ctrl-d and ctrl-u

**Comment by jadtl** [comment]
> I am on MacOs with latest MacOs, latest Alcritty terminal, latest NeoVim. The behavior that OP reported is happening for me now. It happens when either scrolling with mouse or ctrl-d and ctrl-u

I had the same issue and after running `:checkhealth`, `treesitter` was complaining about `nvim`'s version, I thought I had the latest version from Homebrew but `nvim` was actually linked to an old version (0.9.0) I had compiled myself instead of Homebrew's.

Now it doesn't happen to me on Homebrew's 0.9.4. What version `nvim -v` reports for you exactly?

**Comment by agluck91** [comment]
I am on 0.9.4. It seems to only happen on really large files actually. 

There are some warnings in my check health, mostly around parsers, but nothing in an error state. I am missing a couple of executables.
